### ISSUE TITLE
Support LaTeXStrings.jl?

### ISSUE THREAD

**Issue by schneiderfelipe** [JuliaGraphics/Luxor.jl#124]
Hi guys,

Are there any plans on supporting [LaTeXStrings.jl](https://github.com/stevengj/LaTeXStrings.jl)? If not, is it possible to insert an image generated elsewhere in a Luxor picture? Thanks!

**Comment by cormullion** [comment]
Yes, I noticed that Wijunia's Javis.jl supports LATEX. It involves installing node (`npm`) and there's some tricky code involved. It's certainly possible!

You can only insert PNG images into a drawing, but that might do the trick...?

**Comment by schneiderfelipe** [comment]
> Yes, I noticed that Wikunia's Javis.jl supports LATEX. It involves installing node (`npm`) and there's some tricky code involved. It's certainly possible!
> 
> You can only insert PNG images into a drawing, but that might do the trick...?

I think so! As long as the generated PNG has enough resolution (and is transparent?), that would be fine I think.

**Comment by cormullion** [comment]
<img width="1214" alt="Screenshot 2020-12-25 at 14 36 55" src="https://user-images.githubusercontent.com/52289/103137202-b3d8ea00-46be-11eb-984c-5381c519a5bc.png">

The equation PNG is black on transparent background, 854 by 1764 scaled by 0.2, and looks reasonable...

**Comment by schneiderfelipe** [comment]
> The equation PNG is black on transparent background, 854 by 1764 scaled by 0.2, and looks reasonable...

That looks awesome! (By the way, Merry Christmas!) I recently learned that Latexify.jl can generate PDFs from LaTeXStrings (this requires a LaTeX install). I'll make a PR soon to support PNG generation (see korsbo/Latexify.jl#132). Could this be used to somehow automate things on the Luxor.jl side?

**Comment by schneiderfelipe** [comment]
With korsbo/Latexify.jl#133 being merged, the next release of `Latexify.jl` will allow rendering of SVG files as follows:

```julia
using Latexify

tex = L"\frac{1}{\sigma \sqrt{2 \pi}} e^{-\left( \frac{x - \mu}{2 \sigma} \right)^2}"
fname = tempname()  # path to the output file
tex |> s -> render(s, MIME("image/svg"), name=fname)
```

(`MIME("image/png")` also works!)

This could be fed into `readsvg` (#130). Better yet, overloading `text(s::LaTeXString, ...)` (or `latex(...)` if overloading `text(...)` seems like a bad idea?) could automate all this 😃.

**Comment by hustf** [comment]
I spent a day trying to get latex rendering to work on windows with Luxor. 

Some aspects that makes this complicated:
*  LaTeX is much more than formula rendering 
*  MathML, though Julia-expression-friendly, is not used
*  MathJax is for browsers and stylesheet-dependant
*  Glib ecosystem, security when porting to windows and mac
*  node.js dependency explosion
*  Web layout: svg layout is often incomplete without the frames, stylesheets and scripts. Does the viewing window define size?
*  Paper  layout
*  Units of measurement (svg has px, in, cm, mm, pt, pc, em, ex, ch, rem, vw, vh, vmin, vmax, %)

In short, don't get too involved in rendering maths! And when you do, don't link anything! And when you do, don't try for a general solution! And when you do, don't make it perfect.

Currently, Rsvg gives 'incorrect' width and height when reading some svg files. But that's easy for the user to adjust.

Currently, Latexify creates wonderful latex strings, but doesn't create files on all systems. But hacks exist.

I think the current solution in Luxor master is good, as it is not perfect.

**Comment by cormullion** [comment]
Thanks for your efforts, @hustf! Not being in academia, I'm not an expert in LATEX so I'm very grateful for your investigations.

**Comment by hustf** [comment]
Thanks yourself, you're one of the great benefactors to the Julia community and I defininitely had you pictured as a professor. I don't know LaTeX, but as a fan of mechanical engineering 'hand' calculations, I had a little go at starting something similar to MatchCad, but in Julia. It's too large of an open source project, of course. I just wish MathML would have won that standards battle. @korsbo's long battle with latexifying Julia functions just might have been easier, but I'm really grateful for what @latexify does with simple functions now.

**Comment by davibarreira** [comment]
Is there interest in supporting LaTeX in Luxor? I think it could be done with MathTexEngine.jl. But then, it would need to add it as a dependency. If there is interest, I can take a look on how to do it cause I'm working on it myself right now.

**Comment by hustf** [comment]
One vote against dependencies from another user. I feel having a dependency-heavy add-on to Luxor gives the flexibility we need. I use place_image to fit latex equations on a Luxor drawing. Plots are just as useful to put on a Luxor drawing, but nobody is suggesting to add that dependency.

**Comment by davibarreira** [comment]
Yeah, I'm also reticent on adding. I like how fast Luxor is with pre-compilation. @hustf , do you have a helper function to do it?

**Comment by davibarreira** [comment]
Is there a way to perhaps provide the functionality, but leave to the user to import MathTexEngine.jl  without making it a dependency?

**Comment by davibarreira** [comment]
In case this is not implemented. Here is a bit of code I wrote that works. It still need some tweaking in order to properly align, but it's not biggy:
```julia
using Luxor
using LaTeXStrings
using MathTeXEngine

d = Drawing(600,100,:svg)
origin()
fontsize(20)

sentence = generate_tex_elements(L"\frac{\beta_2+10+10\cdot \alpha}{2} \sum^{n}_{i=1} \alpha^2 \int_{-\infty}^\infty x dx \in \bbR")
textsize = get_fontsize()
for text in sentence
    if text[1] isa TeXChar
        @layer begin
            fontface(text[1].font.family_name)
            fontsize(textsize*text[3])
            Luxor.text(string(text[1].char), O + Point(text[2]...)*textsize*(1,-1))
        end
    elseif text[1] isa HLine
        @layer begin
            pointstart = O+Point(text[2]...)*textsize*(1,-1)
            pointend = pointstart + Point(text[1].width,0)*textsize
            setline(1*text[1].thickness*textsize)
            line(pointstart, pointend,:stroke)
        end
    end
end
finish()
d
```

**Comment by hustf** [comment]
Hi, I had a brief look at `MathTexEngine.jl`, and it seems both lightweight and promising. I have not used it. In my own dependency-heavy package, I recently dropped ForwardDiff as a dependency, and to get things working again with minumum effort, the examples are now commented out. A primitive example of latexified expressions is in [test36.png](https://github.com/hustf/MechanicalSketch.jl/blob/master/test/test_36.png), made by 'test36.jl':26 and so forth. Package code which you might be interested in copying is in 'src/latex.jl', 'macros.jl' and 'place_image.jl' in the same repo. Good luck with your approach!

**Comment by davibarreira** [comment]
@cormullion , I've implemented much of the  functionality. There are just some small issues related to vertical alignment. But it's already quite usable. Let me know if I should submit a PR.

**Comment by cormullion** [comment]
Using Requires.jl (https://github.com/JuliaPackaging/Requires.jl) “magically makes it easy” to add conditional dependencies. It might be just the thing here! Would be great if we could get this going. 😂

**Comment by TheCedarPrince** [comment]
Hey @davibarreira and @cormullion ,

I just tried out the code snippet here: https://github.com/JuliaGraphics/Luxor.jl/issues/124#issuecomment-983127493
And it seemingly failed to work for me:

![image](https://user-images.githubusercontent.com/29561456/147496625-63ae37d0-f185-4e80-98d5-56187dfb21cb.png)

I didn't receive any traceback or error.
Is there any reason why this may be happening?

**Comment by davibarreira** [comment]
> Using Requires.jl (https://github.com/JuliaPackaging/Requires.jl) “magically makes it easy” to add conditional dependencies. It might be just the thing here! Would be great if we could get this going. joy

Nice, I'll look into it.

@TheCedarPrince , give me a minute, I'll test the code to see if I can replicate. I was using a notebook, so maybe I had something else in the kernel and didn't realize.

**Comment by davibarreira** [comment]
I tested here, and it worked. Did you try running this in a notebook? What are you using to visualize the svg? I just checked the generated svg, and it turns the letters into paths, so I don't think you need to install it. If running in the notebook does not work, then maybe try installing CMU.

**Comment by TheCedarPrince** [comment]
Hey @davibarreira -

I am fiddling wiht this today, and realized that the issue is not with notebooks or what have you but rather the issue is within MathTexEngine itself. 
Once I installed the CMU fonts to my system via Lyx fonts, it rendered LaTeX properly.
Somehow, there is a sourcing issue with MathTexEngine's fonts as it _should_ be sourced and working.
Will keep investigating.

**Comment by TheCedarPrince** [comment]
Ok. 
After a couple hours of scrutiny, I have determined the crux of the issue with MathTeXEngine and Luxor compatibility.
Long story short, fonts are extremely evil and must be configured generally on a per system basis as Luxor (via Cairo) utilizes a sort of `fontconfig` tool of its own and looks on a system basis to where fonts are stored.
It is non-trivial to add TTF based fonts (or otherwise) via Cairo.jl and MathTeXEngine so frankly I would just suggest it up to the user to install the ModernCMU fonts on their computer.
Otherwise, @davibarreira , your solution works perfectly:

![image](https://user-images.githubusercontent.com/29561456/147519372-72c15d18-a807-4fc9-bcc2-331a40e9d1ed.png)

Minus this somewhat strange behavior in the center of the diagram.
Somehow, the LaTeX is following the slope of the line!

**Comment by davibarreira** [comment]
I havent tried out writing it in a line. Can you share the code?

Em seg., 27 de dez. de 2021 23:00, TheCedarPrince ***@***.***>
escreveu:

**Comment by cormullion** [comment]
>Long story short, fonts are extremely evil

I think that’s a fair summary! 😂 Finding a cross-platform solution to any font-related issue is always difficult or impossible.

**Comment by TheCedarPrince** [comment]
@davibarreira , here is a [gist](https://gist.github.com/TheCedarPrince/11965e9aeb8c2cd6ef451b1df7ba2bd7) of what happens.
If you have questions about the gist, please ask.
As this is just me rapidly prototyping ideas, I did add some documentation but am happy to jump on a call with you at some point to trace where things are happening.
To me, it seems like since each symbol generated by your approach is treated separately, somehow, alignment is getting screwed up.

You can see the difference between a standard Luxor `text` block and the LaTeX rendered `text` in this image:

![image](https://user-images.githubusercontent.com/29561456/147597354-c133d411-382c-4129-9007-a138c5e1c752.png)

> I think that’s a fair summary! joy Finding a cross-platform solution to any font-related issue is always difficult or impossible.

Haha, glad you agreed @cormullion !
Didn't realize how difficult it was until I tinkered with it for a couple hours yesterday.

Also, once upon a time @Wikunia and I did try using Requires.jl but we found it quite cumbersome and eventually gave up on the experiment.
However, it may be different for your use case.
We however, kept running into the world age problem - which was the first time my beginner Julia eyes had even seen such a problem and thought we had ripped a hole in the space time continuum.

**Comment by cormullion** [comment]
@TheCedarPrince Looks cool! - but I haven't seen text set on a diagonal line before - is it an essential part of the visualization tool you're making? Since the characters are all different widths, I think it might look odd anyway. 

In other graphic environments (eg drawing apps), you would probably "shear" text, with a shear transform applied to each character. This maintains the text's baseline:

<img width="417" alt="Screenshot 2021-12-28 at 19 15 41" src="https://user-images.githubusercontent.com/52289/147599508-5ede813b-741a-4005-a4fd-7803d407cf42.png">
 
Re Requires:  thought that since over 2000 Julia packages depend on Requires.jl it must work OK - perhaps it doesn't, I've not tried it yet... 🤔

**Comment by TheCedarPrince** [comment]
Ope! 
Apologies I was not clear enough @cormullion and @davibarreira !
I am trying to _prevent_ the diagonal text from showing up.
Instead, I want it to read on a horizontal axis from left to right like the other non-LaTeX font.

Ah, well, I could be completely wrong with Requires.jl. 
We had an issue about it somewhere on Javis.
I will see if I can tag you in it and show you what we tried.

**Comment by cormullion** [comment]
Haha, I got the wrong idea - makes sense now. 😀

**Comment by davibarreira** [comment]
I've got the flu, so I've been absent these days. @TheCedarPrince , what is the function that you are using to write the text with respect to this diagonal line? The code snippet I posted is not complete, for example, it does not have the rotating functionality.
I've already implemented it, and it'll be available in the PR.

**Comment by TheCedarPrince** [comment]
Oh no - I hope you are feeling better @davibarreira !

This is the function I use for drawing text: https://gist.github.com/TheCedarPrince/11965e9aeb8c2cd6ef451b1df7ba2bd7#file-latex_error-jl-L47

And this is the function that eventually calls it after calculating the angle needed: https://gist.github.com/TheCedarPrince/11965e9aeb8c2cd6ef451b1df7ba2bd7#file-latex_error-jl-L170

If you open the PR, happy to test it out fully.
As you said, as it was not complete earlier, perhaps that is why this behavior is existing.

**Comment by davibarreira** [comment]
PR opened.
It has the full functionality (alignment and rotation) which are missing in the snippet I shared here. Anyone interested, can take a look. All the code is  inside `latex.jl` in the PR, and can be easily copied in case you wish to use it by itself.

**Comment by davibarreira** [comment]
@TheCedarPrince , I'd be glad if you could check whether the behavior is correct in your diagonal line.

**Comment by davibarreira** [comment]
@cormullion , I'm guessing the new features will only be available once Luxor get's a new release, right? I've updated in my personal env here, but I'm still not seeing the changes.

**Comment by cormullion** [comment]
Yes, the next release should be this week perhaps, given the great improvements you've made! Thanks again.

**Comment by davibarreira** [comment]
Glad I could help! I love Luxor, and it's great to help it grow. :)

**Comment by TheCedarPrince** [comment]
@davibarreira and @cormullion !

Great news!
This works perfectly for me!

![image](https://user-images.githubusercontent.com/29561456/147894739-a10f5a46-9aad-4708-ae03-7977d93e15c2.png)

Right now I am speaking with @kolaru to see if they can potentially add some sort of functionality or warning to a user that if CMU fonts are not installed, MathTeXEngine will not work properly.

I saw that in #191 there were docs added for LaTeXStrings.
I'll open a mini-PR to put a small disclaimer in about the fonts. 

However, dare I say, is it time to close this issue?

**Comment by davibarreira** [comment]
I think it can be closed. I added a note on the docs about installing the fonts :D

**Comment by TheCedarPrince** [comment]
Ope! :man_facepalming: 
Alright! 
@cormullion , at your discretion I suppose!

**Comment by cormullion** [comment]
It's looking pretty good! I think there's no way of automatically installing fonts reliably on all three platforms, so it will probably be always up to the user to do the necessary font installation work. 

It wouldn't be proper LaTeX if there weren't font issues as well... :)
